File: cpp/client/client_data.cc
#include <webcface/client_data.h>

namespace WebCFace {
template <typename T>
void ClientData::SyncDataStore2<T>::setSend(const std::string &name,
                                            const T &data) {
    std::lock_guard lock(mtx);
    data_send[name] = data;
    data_recv[self_member_name][name] = data; // 送信後に自分の値を参照する用
}
template <typename T>
void ClientData::SyncDataStore2<T>::setHidden(const std::string &name,
                                              bool is_hidden) {
    std::lock_guard lock(mtx);
    data_send_hidden[name] = is_hidden;
}
template <typename T>
bool ClientData::SyncDataStore2<T>::isHidden(const std::string &name) {
    std::lock_guard lock(mtx);
    auto h = data_send_hidden.find(name);
    return h != data_send_hidden.end() && h->second == true;
}

template <typename T>
void ClientData::SyncDataStore2<T>::setRecv(const std::string &from,
                                            const std::string &name,
                                            const T &data) {
    std::lock_guard lock(mtx);
    data_recv[from][name] = data;
}
template <typename T>
void ClientData::SyncDataStore1<T>::setRecv(const std::string &member,
                                            const T &data) {
    std::lock_guard lock(mtx);
    data_recv[member] = data;
}
template <typename T>
template <typename U>
    requires std::same_as<T, std::vector<U>>
void ClientData::SyncDataStore1<T>::addRecv(const std::string &member,
                                            const U &data) {
    std::lock_guard lock(mtx);
    data_recv[member].push_back(data);
}

template <typename T>
std::vector<std::string> ClientData::SyncDataStore2<T>::getMembers() {
    std::lock_guard lock(mtx);
    std::vector<std::string> k;
    for (const auto &r : entry) {
        k.push_back(r.first);
    }
    return k;
}
template <typename T>
std::vector<std::string>
ClientData::SyncDataStore2<T>::getEntry(const std::string &name) {
    std::lock_guard lock(mtx);
    auto e = entry.find(name);
    if (e != entry.end()) {
        return e->second;
    } else {
        return std::vector<std::string>{};
    }
}
template <typename T>
void ClientData::SyncDataStore2<T>::setEntry(const std::string &from) {
    std::lock_guard lock(mtx);
    entry.emplace(std::make_pair(from, std::vector<std::string>{}));
}
template <typename T>
void ClientData::SyncDataStore2<T>::setEntry(const std::string &from,
                                             const std::string &e) {
    std::lock_guard lock(mtx);
    entry[from].push_back(e);
}

template <typename T>
std::optional<T>
ClientData::SyncDataStore2<T>::getRecv(const std::string &from,
                                       const std::string &name) {
    std::lock_guard lock(mtx);
    if (!isSelf(from) && req[from][name] <= 0) {
        unsigned int max_req = 0;
        for (const auto &r : req) {
            for (const auto &r2 : r.second) {
                if (r2.second > max_req) {
                    max_req = r2.second;
                }
            }
        }
        req[from][name] = max_req + 1;
        req_send[from][name] = max_req + 1;
    }
    auto s_it = data_recv.find(from);
    if (s_it != data_recv.end()) {
        auto it = s_it->second.find(name);
        if (it != s_it->second.end()) {
            return it->second;
        }
    }
    return std::nullopt;
}
template <typename T>
std::optional<T>
ClientData::SyncDataStore1<T>::getRecv(const std::string &member) {
    std::lock_guard lock(mtx);
    if (!isSelf(member) && req[member] == false) {
        req[member] = true;
        req_send[member] = true;
    }
    auto s_it = data_recv.find(member);
    if (s_it != data_recv.end()) {
        return s_it->second;
    }
    return std::nullopt;
}
template <typename T>
void ClientData::SyncDataStore2<T>::unsetRecv(const std::string &from,
                                              const std::string &name) {
    std::lock_guard lock(mtx);
    if (!isSelf(from) && req[from][name] > 0) {
        req[from].erase(name);
        req_send[from][name] = 0;
    }
    if (data_recv.count(from) && data_recv.at(from).count(name)) {
        data_recv.at(from).erase(name);
    }
}
template <typename T>
std::pair<std::string, std::string>
ClientData::SyncDataStore2<T>::getReq(unsigned int req_id) {
    std::lock_guard lock(mtx);
    for (const auto &r : req) {
        for (const auto &r2 : r.second) {
            if (r2.second == req_id) {
                return std::make_pair(r.first, r2.first);
            }
        }
    }
    return std::make_pair("", "");
}

template <typename T>
std::unordered_map<std::string, T>
ClientData::SyncDataStore2<T>::transferSend(bool is_first) {
    std::lock_guard lock(mtx);
    if (is_first) {
        data_send.clear();
        return data_send_prev = data_recv[self_member_name];
    } else {
        data_send_prev = data_send;
        return std::move(data_send);
    }
}
template <typename T>
std::unordered_map<std::string, T>
ClientData::SyncDataStore2<T>::getSendPrev(bool is_first) {
    std::lock_guard lock(mtx);
    if (is_first) {
        return std::unordered_map<std::string, T>{};
    } else {
        return data_send_prev;
    }
}
template <typename T>
std::unordered_map<std::string, std::unordered_map<std::string, unsigned int>>
ClientData::SyncDataStore2<T>::transferReq(bool is_first) {
    std::lock_guard lock(mtx);
    if (is_first) {
        req_send.clear();
        return req;
    } else {
        return std::move(req_send);
    }
}
template <typename T>
std::unordered_map<std::string, bool>
ClientData::SyncDataStore1<T>::transferReq(bool is_first) {
    std::lock_guard lock(mtx);
    if (is_first) {
        req_send.clear();
        return req;
    } else {
        return std::move(req_send);
    }
}

template class ClientData::SyncDataStore2<double>;
template class ClientData::SyncDataStore2<std::string>;
template class ClientData::SyncDataStore2<FuncInfo>;
template class ClientData::SyncDataStore2<std::vector<ViewComponentBase>>;
template class ClientData::SyncDataStore1<std::vector<LogLine>>;
template void
ClientData::SyncDataStore1<std::vector<LogLine>>::addRecv(const std::string &,
                                                          const LogLine &);
template class ClientData::SyncDataStore1<
    std::chrono::system_clock::time_point>;

AsyncFuncResult &
ClientData::FuncResultStore::addResult(const std::string &caller,
                                       const Field &base) {
    std::lock_guard lock(mtx);
    unsigned int caller_id = results.size();
    results.push_back(AsyncFuncResult{caller_id, caller, base});
    return results.back();
}
AsyncFuncResult &ClientData::FuncResultStore::getResult(int caller_id) {
    std::lock_guard lock(mtx);
    return results.at(caller_id);
}

std::string ClientData::getMemberNameFromId(unsigned int id) const {
    for (const auto &it : member_ids) {
        if (it.second == id) {
            return it.first;
        }
    }
    return "";
}
unsigned int ClientData::getMemberIdFromName(const std::string &name) const {
    auto it = member_ids.find(name);
    if (it != member_ids.end()) {
        return it->second;
    }
    return 0;
}
} // namespace WebCFace

File: client_js/src/clientData.ts
import { Val, FuncInfo, AsyncFuncResult } from "./func.js";
import { EventEmitter } from "eventemitter3";
import { LogLine } from "./logger.js";
import * as Message from "./message.js";
import { FieldBase, Field } from "./field.js";

export class ClientData {
  selfMemberName: string;
  valueStore: SyncDataStore2<number[]>;
  textStore: SyncDataStore2<string>;
  funcStore: SyncDataStore2<FuncInfo>;
  viewStore: SyncDataStore2<Message.ViewComponent[]>;
  logStore: SyncDataStore1<LogLine[]>;
  syncTimeStore: SyncDataStore1<Date>;
  funcResultStore: FuncResultStore;
  memberIds: Map<string, number>;
  callFunc: (r: AsyncFuncResult, b: FieldBase, args: Val[]) => void;
  eventEmitter: EventEmitter;
  logQueue: LogLine[];
  constructor(
    name: string,
    callFunc: (r: AsyncFuncResult, b: FieldBase, args: Val[]) => void
  ) {
    this.selfMemberName = name;
    this.valueStore = new SyncDataStore2<number[]>(name);
    this.textStore = new SyncDataStore2<string>(name);
    this.funcStore = new SyncDataStore2<FuncInfo>(name);
    this.viewStore = new SyncDataStore2<Message.ViewComponent[]>(name);
    this.logStore = new SyncDataStore1<LogLine[]>(name);
    this.syncTimeStore = new SyncDataStore1<Date>(name);
    this.funcResultStore = new FuncResultStore();
    this.memberIds = new Map<string, number>();
    this.callFunc = callFunc;
    this.eventEmitter = new EventEmitter();
    this.logQueue = [];
  }
  isSelf(member: string) {
    return this.selfMemberName === member;
  }
  getMemberNameFromId(id: number) {
    for (const [n, i] of this.memberIds.entries()) {
      if (i === id) {
        return n;
      }
    }
    return "";
  }
  getMemberIdFromName(name: string) {
    return this.memberIds.get(name) || 0;
  }
}

export class SyncDataStore2<T> {
  dataSend: Map<string, T>;
  dataSendPrev: Map<string, T>;
  dataSendHidden: Map<string, boolean>;
  dataRecv: Map<string, Map<string, T>>;
  entry: Map<string, string[]>;
  req: Map<string, Map<string, number>>;
  reqSend: Map<string, Map<string, number>>;
  selfMemberName: string;
  constructor(name: string) {
    this.selfMemberName = name;
    this.dataSend = new Map();
    this.dataSendPrev = new Map();
    this.dataSendHidden = new Map();
    this.dataRecv = new Map();
    this.entry = new Map();
    this.req = new Map();
    this.reqSend = new Map();
  }
  isSelf(member: string) {
    return this.selfMemberName === member;
  }
  //! 送信するデータをdata_sendとdata_recv[self_member_name]にセット
  setSend(field: string, data: T) {
    this.dataSend.set(field, data);
    this.setRecv(this.selfMemberName, field, data);
  }
  setHidden(field: string, isHidden: boolean) {
    this.dataSendHidden.set(field, isHidden);
  }
  isHidden(field: string) {
    return this.dataSendHidden.get(field) == true;
  }
  //! 受信したデータをdata_recvにセット
  setRecv(member: string, field: string, data: T) {
    const m = this.dataRecv.get(member);
    if (m) {
      m.set(field, data);
    } else {
      this.dataRecv.set(member, new Map([[field, data]]));
    }
  }
  getMaxReq() {
    let maxReq = 0;
    for (const [rm, r] of this.req.entries()) {
      for (const [rf, ri] of r.entries()) {
        if (ri > maxReq) {
          maxReq = ri;
        }
      }
    }
    return maxReq;
  }
  //! data_recvからデータを返す or なければreq,req_sendをtrueにセット
  getRecv(member: string, field: string) {
    if (!this.isSelf(member) && !this.req.get(member)?.get(field)) {
      const m = this.req.get(member);
      const newReq = this.getMaxReq() + 1;
      if (m) {
        m.set(field, newReq);
      } else {
        this.req.set(member, new Map([[field, newReq]]));
      }
      const m2 = this.reqSend.get(member);
      if (m2) {
        m2.set(field, newReq);
      } else {
        this.reqSend.set(member, new Map([[field, newReq]]));
      }
    }
    const m = this.dataRecv.get(member)?.get(field);
    if (m != undefined) {
      return m;
    }
    return null;
  }
  //! data_recvからデータを削除, req,req_sendをfalseにする
  unsetRecv(member: string, field: string) {
    if (!this.isSelf(member) && !!this.req.get(member)?.get(field)) {
      this.req.get(member)?.set(field, 0);
      const m2 = this.reqSend.get(member);
      if (m2) {
        m2.set(field, 0);
      } else {
        this.reqSend.set(member, new Map([[field, 0]]));
      }
    }
    this.dataRecv.get(member)?.delete(field);
  }
  //! member名のりすとを取得(entryから)
  getMembers() {
    return Array.from(this.entry.keys());
  }
  //! entryを取得
  getEntry(member: string) {
    return this.entry.get(member) || [];
  }
  //! entryにmember名のみ追加
  addMember(member: string) {
    this.entry.set(member, []);
  }
  //! 受信したentryを追加
  setEntry(member: string, e: string) {
    this.entry.set(member, (this.getEntry(member) || []).concat([e]));
  }
  //! data_sendを返し、data_sendをクリア
  transferSend(isFirst: boolean) {
    if (isFirst) {
      this.dataSend = new Map();
      // dataSendPrevはdataRecvが書き換えられても影響しないようコピーする
      this.dataSendPrev = new Map();
      const dataCurrent =
        this.dataRecv.get(this.selfMemberName) || new Map<string, T>();
      for (const [k, v] of dataCurrent.entries()) {
        this.dataSendPrev.set(k, v);
      }
      return dataCurrent;
    } else {
      const s = this.dataSend;
      this.dataSendPrev = s;
      this.dataSend = new Map();
      return s;
    }
  }
  getSendPrev(isFirst: boolean) {
    if (isFirst) {
      return new Map<string, T>();
    } else {
      return this.dataSendPrev;
    }
  }
  //! req_sendを返し、req_sendをクリア
  transferReq(isFirst: boolean) {
    if (isFirst) {
      this.reqSend = new Map();
      return this.req;
    } else {
      const r = this.reqSend;
      this.reqSend = new Map();
      return r;
    }
  }
  getReq(i: number, subField: string) {
    for (const [rm, r] of this.req.entries()) {
      for (const [rf, ri] of r.entries()) {
        if (ri == i) {
          return [rm, subField !== "" ? rf + "." + subField : rf];
        }
      }
    }
    return ["", ""];
  }
}

export class SyncDataStore1<T> {
  dataRecv: Map<string, T>;
  req: Map<string, boolean>;
  reqSend: Map<string, boolean>;
  selfMemberName: string;
  constructor(name: string) {
    this.selfMemberName = name;
    this.dataRecv = new Map();
    this.req = new Map();
    this.reqSend = new Map();
  }
  isSelf(member: string) {
    return this.selfMemberName === member;
  }
  setRecv(member: string, data: T) {
    this.dataRecv.set(member, data);
  }
  getRecv(member: string) {
    if (!this.isSelf(member) && this.req.get(member) !== true) {
      this.req.set(member, true);
      this.reqSend.set(member, true);
    }
    const m = this.dataRecv.get(member);
    if (m != undefined) {
      return m;
    }
    return null;
  }
  transferReq(isFirst: boolean) {
    if (isFirst) {
      this.reqSend = new Map();
      return this.req;
    } else {
      const r = this.reqSend;
      this.reqSend = new Map();
      return r;
    }
  }
}

export class FuncResultStore {
  results: AsyncFuncResult[] = [];
  addResult(caller: string, base: Field) {
    const callerId = this.results.length;
    this.results.push(new AsyncFuncResult(callerId, caller, base));
    return this.results[callerId];
  }
  getResult(callerId: number) {
    return this.results[callerId];
  }
}
